dashboard/app/controllers/activities_controller.rb

```diff
@@ -15,52 +15,92 @@ class ActivitiesController < ApplicationController
+  # - hint_view_request_url
+  # - hint_view_requests
+  # - level_id
+  # - level_source
+  # - level_source_id
+  # - message
+  # - phone_share_url
+  # - script_id
+  #
   def milestone
-    # TODO: do we use the :result and :testResult params for the same thing?
-    solved = ('true' == params[:result])
-    if params[:script_level_id]
-      @script_level = ScriptLevel.cache_find(params[:script_level_id].to_i)
-      @level = @script_level.level
-    elsif params[:level_id]
-      # TODO: do we need a cache_find for Level like we have for ScriptLevel?
-      @level = Level.find(params[:level_id].to_i)
-    end
-
-    if params[:program]
-      begin
-        share_failure = find_share_failure(params[:program])
-      rescue OpenURI::HTTPError => share_checking_error
-        # If WebPurify fails, the program will be allowed
-      end
-
-      unless share_failure
-        @level_source = LevelSource.find_identical_or_create(@level, params[:program])
-        slog(tag: 'share_checking_error', error: "#{share_checking_error}", level_source_id: @level_source.id) if share_checking_error
-      end
-    end
-
-    if params[:lines]
-      params[:lines] = params[:lines].to_i
-      params[:lines] = 0 if params[:lines] < MIN_LINES_OF_CODE
-      params[:lines] = MAX_LINES_OF_CODE if params[:lines] > MAX_LINES_OF_CODE
-    end
-
-    # Store the image only if the image is set, and the image has not been saved
-    if params[:image] && @level_source.try(:id)
-      @level_source_image = LevelSourceImage.find_by(level_source_id: @level_source.id)
-      unless @level_source_image
-        @level_source_image = LevelSourceImage.new(level_source_id: @level_source.id)
-        unless @level_source_image.save_to_s3(Base64.decode64(params[:image]))
-          @level_source_image = nil
-        end
-      end
+    # Clean up and pull out params
+    save_to_gallery = params[:save_to_gallery] == 'true'
+    solved = params[:result] == 'true'
+    test_result = nil_or_int(params[:testResult])
+    attempt = nil_or_int(params[:attempt])
+    script_level_id = nil_or_int(params[:script_level_id])
+    level_id = nil_or_int(params[:level_id])
+    image = params[:image]
+    lines = params[:lines].to_i
+    program = params[:program]
+
+    # Figure out what level and script_level we're at
+    script_level = ScriptLevel.cache_find(script_level_id.to_i) if script_level_id
+    if script_level
+      level = script_level.level
+    elsif level_id
+      level = Level.find(level_id.to_i)
     end
 
-    @new_level_completed = false
-    if current_user
-      track_progress_for_user if @script_level
-    else
-      track_progress_in_session
+    result = track_level_solution(
+      level: level,
+      program: program,
+      image: image,
+    )
+
+    level_source = result[:level_source]
+    level_source_image = result[:level_source_image]
+    share_failure = result[:share_failure]
+
+    lines = [[0, lines.to_i].max, MAX_LINES_OF_CODE].min if lines
```

Comment:
the 0 here should use the MIN_LINES_OF_CODE constant, or if unused it could be removed.